File: app/src/main/java/io/github/takusan23/Kaisendon/MarqueeTextView.java
package io.github.takusan23.Kaisendon;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.widget.TextView;

public class MarqueeTextView extends TextView {

    private boolean mIsMarqueeStarted;

    public MarqueeTextView(Context context) {
        this(context, null);
    }

    public MarqueeTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MarqueeTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        setEllipsize(TextUtils.TruncateAt.MARQUEE);
        setMarqueeRepeatLimit(-1);
        setFocusable(true);
        setFocusableInTouchMode(true);
        setSingleLine(true);
    }

    @Override
    public boolean isSelected() {
        // Always return true. because marquee animation stop when return false.
        return true;
    }

    public void startMarquee() {
        if (mIsMarqueeStarted) {
            return;
        }
        mIsMarqueeStarted = true;
        setEllipsize(TextUtils.TruncateAt.MARQUEE);
        invalidate();
    }

    public void stopMarquee() {
        if (!mIsMarqueeStarted) {
            return;
        }
        mIsMarqueeStarted = false;
        setEllipsize(TextUtils.TruncateAt.END);
        invalidate();
    }
}

File: app/src/main/java/io/github/takusan23/Kaisendon/MarqueeTextView.kt
package io.github.takusan23.Kaisendon

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.widget.TextView

class MarqueeTextView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : TextView(context, attrs, defStyleAttr) {

    private var mIsMarqueeStarted: Boolean = false

    init {

        ellipsize = TextUtils.TruncateAt.MARQUEE
        marqueeRepeatLimit = -1
        isFocusable = true
        isFocusableInTouchMode = true
        isSingleLine = true
    }

    override fun isSelected(): Boolean {
        // Always return true. because marquee animation stop when return false.
        return true
    }

    fun startMarquee() {
        if (mIsMarqueeStarted) {
            return
        }
        mIsMarqueeStarted = true
        ellipsize = TextUtils.TruncateAt.MARQUEE
        invalidate()
    }

    fun stopMarquee() {
        if (!mIsMarqueeStarted) {
            return
        }
        mIsMarqueeStarted = false
        ellipsize = TextUtils.TruncateAt.END
        invalidate()
    }
}